src/lib/orion_websocket/wsNotify.cpp

```diff
@@ -0,0 +1,51 @@
+#include "wsNotify.h"
+#include "constants.h"
+#include "parser.h"
+
+#include <libwebsockets.h>
+#include <cstring>
+
+
+std::map<std::string, lws*>client;
+
+void addSender(const std::string &subId, lws *ws)
+{
+  client[subId] = ws;
+}
+
+void removeSenders(const std::vector<std::string>& subIds)
+{
+  for (unsigned i = 0; i < subIds.size(); ++i)
+  {
+    client.erase(subIds[i]);
+  }
+}
+
+int sendNotifyContextRequestWs
+(
+  const std::string& subId,
+  const std::map<std::string, std::string>& headers,
+  const std::string& data
+)
+{
+  const char *msg = ws_parser_notify(subId, headers, data);
+
+  std::map<std::string, lws*>::iterator it = client.find(subId);
+
+  if (it == client.end())
+    return 1;
+
+  size_t size = strlen(msg) +
+                LWS_SEND_BUFFER_PRE_PADDING +
+                LWS_SEND_BUFFER_POST_PADDING;
+
+  unsigned char *buff = (unsigned char *) malloc(size);
+  unsigned char *p = &buff[LWS_SEND_BUFFER_PRE_PADDING];
+  sprintf((char *)p, "%s", msg);
+  int result = lws_write(it->second, p, strlen(msg), LWS_WRITE_TEXT);
+
+  free((char *)msg);
+  free(buff);
+
+  return result < 0;
```

Comment:
I can.

Are we guaranteed that lws_write always sends the entire buffer?

Look at the 'normal' write function.
It returns the number of bytes written. It can perfectly be called with a buffer of 1024 bytes and return an 'ok', saying that "I have sent 259 bytes". What you do then is to call *write* again, with the remaining 765 bytes, and so on, in a loop until the entire buffer has been sent.

So, my question is:  do we know for sure that lws_write sends the entire buffer?